File: src/data/courses/dsa/merge-sort/sorting-comparisons/time-complexity-analysis/code/java/TimeComplexityAnalysis.java
// Merge Sort: O(n log n) always
void mergeSort(int[] arr, int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;
        mergeSort(arr, l, m);        // T(n/2)
        mergeSort(arr, m + 1, r);    // T(n/2)
        merge(arr, l, m, r);         // O(n)
    }
}
// Recurrence: T(n) = 2T(n/2) + O(n) = O(n log n)

// Quick Sort: O(n log n) average, O(n²) worst
void quickSort(int[] arr, int l, int r) {
    if (l < r) {
        int p = partition(arr, l, r);  // O(n)
        quickSort(arr, l, p - 1);      // T(n/2) avg
        quickSort(arr, p + 1, r);      // T(n/2) avg
    }
}
// Best/Avg: T(n) = 2T(n/2) + O(n) = O(n log n)
// Worst: T(n) = T(n-1) + O(n) = O(n²)

File: src/data/courses/dsa/merge-sort/sorting-comparisons/time-complexity-analysis/code/python/time_complexity_analysis.py
# Merge Sort: O(n log n)
def merge_sort(arr, l, r):
    if l < r:
        m = l + (r - l) // 2
        merge_sort(arr, l, m)      # T(n/2)
        merge_sort(arr, m + 1, r)  # T(n/2)
        merge(arr, l, m, r)        # O(n)
# T(n) = 2T(n/2) + O(n) = O(n log n)

# Quick Sort: O(n log n) avg, O(n²) worst
def quick_sort(arr, l, r):
    if l < r:
        p = partition(arr, l, r)
        quick_sort(arr, l, p - 1)
        quick_sort(arr, p + 1, r)